### ISSUE TITLE
Exception taken importing a valid SBOM. Appears to be bom-ref related

### ISSUE THREAD

**Issue by nigellh** [DependencyTrack/dependency-track#3469]
### Current Behavior

According to the docker-compose.yml file the system is running version: '3.7' (It does not allow me to select this version below, so please fix that if possible)

I used Trivy (https://github.com/aquasecurity/trivy) to scan a UBI Container (registry.access.redhat.com/ubi9/ubi:9.3) and generate an SBOM.

The entry for running it and getting a CDX SBOM out is

./trivy image --output ubi_standard_9.3.json --format cyclonedx --report all registry.access.redhat.com/ubi9/ubi:9.3

(I inadvertently typed uib in originally and is the name of the zip below)

I have attached the SBOM below and it is a CDX 1.5 version. I validated it using the CycloneDX-CLI tool and it passed.

I create a project in DT and imported the SBOM and nothing showed. Looking at the logs I get.

```
[36mdtrack-apiserver_1  |[0m 2024-02-15 09:23:12,860 INFO [ProjectResource] Project UBI Standard Trivy : 9.3-1552 created by nhopper
[36mdtrack-apiserver_1  |[0m 2024-02-15 09:23:41,236 INFO [BomUploadProcessingTask] Processing CycloneDX BOM uploaded to project: 344ad184-f1dd-44a7-ba97-5facb6f748d1
[36mdtrack-apiserver_1  |[0m 2024-02-15 09:23:41,252 ERROR [BomUploadProcessingTask] Error while processing bom
[36mdtrack-apiserver_1  |[0m org.cyclonedx.exception.ParseException: Unable to parse BOM from byte array
[36mdtrack-apiserver_1  |[0m 	at org.cyclonedx.parsers.JsonParser.parse(JsonParser.java:70)
[36mdtrack-apiserver_1  |[0m 	at org.dependencytrack.tasks.BomUploadProcessingTask.inform(BomUploadProcessingTask.java:105)
[36mdtrack-apiserver_1  |[0m 	at alpine.event.framework.BaseEventService.lambda$publish$0(BaseEventService.java:101)
[36mdtrack-apiserver_1  |[0m 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
[36mdtrack-apiserver_1  |[0m 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
[36mdtrack-apiserver_1  |[0m 	at java.base/java.lang.Thread.run(Unknown Source)
[36mdtrack-apiserver_1  |[0m Caused by: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot deserialize value of type `java.util.ArrayList<org.cyclonedx.model.Tool>` from Object value (token `JsonToken.START_OBJECT`)
[36mdtrack-apiserver_1  |[0m  at [Source: (byte[])"{
[36mdtrack-apiserver_1  |[0m   "$schema": "http://cyclonedx.org/schema/bom-1.5.schema.json",
[36mdtrack-apiserver_1  |[0m   "bomFormat": "CycloneDX",
[36mdtrack-apiserver_1  |[0m   "specVersion": "1.5",
[36mdtrack-apiserver_1  |[0m   "serialNumber": "urn:uuid:94502407-3c9c-402d-8e90-d0763fc3fe78",
[36mdtrack-apiserver_1  |[0m   "version": 1,
[36mdtrack-apiserver_1  |[0m   "metadata": {
[36mdtrack-apiserver_1  |[0m     "timestamp": "2024-02-15T09:09:27+00:00",
[36mdtrack-apiserver_1  |[0m     "tools": {
[36mdtrack-apiserver_1  |[0m       "components": [
[36mdtrack-apiserver_1  |[0m         {
[36mdtrack-apiserver_1  |[0m           "type": "application",
[36mdtrack-apiserver_1  |[0m           "group": "aquasecurity",
[36mdtrack-apiserver_1  |[0m           "name": "trivy",
[36mdtrack-apiserver_1  |[0m           "version": "0.49.1"
[36mdtrack-apiserver_1  |[0m         }
[36mdtrack-apiserver_1  |[0m       ]
[36mdtrack-apiserver_1  |[0m     },
[36mdtrack-apiserver_1  |[0m     "component": {
[36mdtrack-apiserver_1  |[0m       "bom-ref": "pkg"[truncated 362924 bytes]; line: 9, column: 14] (through reference chain: org.cyclonedx.model.Bom["metadata"]->org.cyclonedx.model.Metadata["tools"])
[36mdtrack-apiserver_1  |[0m 	at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:59)
[36mdtrack-apiserver_1  |[0m 	at com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1746)
[36mdtrack-apiserver_1  |[0m 	at com.fasterxml.jackson.databind.DeserializationContext.handleUnexpectedToken(DeserializationContext.java:1520)
[36mdtrack-apiserver_1  |[0m 	at com.fasterxml.jackson.databind.DeserializationContext.handleUnexpectedToken(DeserializationContext.java:1467)
[36mdtrack-apiserver_1  |[0m 	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.handleNonArray(CollectionDeserializer.java:396)
[36mdtrack-apiserver_1  |[0m 	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:252)
[36mdtrack-apiserver_1  |[0m 	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:28)
[36mdtrack-apiserver_1  |[0m 	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
[36mdtrack-apiserver_1  |[0m 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:314)
[36mdtrack-apiserver_1  |[0m 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
[36mdtrack-apiserver_1  |[0m 	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
[36mdtrack-apiserver_1  |[0m 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:314)
[36mdtrack-apiserver_1  |[0m 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
[36mdtrack-apiserver_1  |[0m 	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:323)
[36mdtrack-apiserver_1  |[0m 	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4730)
[36mdtrack-apiserver_1  |[0m 	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3738)
[36mdtrack-apiserver_1  |[0m 	at org.cyclonedx.parsers.JsonParser.parse(JsonParser.java:68)
[36mdtrack-apiserver_1  |[0m 	... 5 common frames omitted
```

I have searched the issues for this type of information and not found anything so do not know if it has been fixed in a later version.

[uib_standard_9.3.json.zip](https://github.com/DependencyTrack/dependency-track/files/14296533/uib_standard_9.3.json.zip)

### Steps to Reproduce

1. See above

### Expected Behavior

SBOM to be imported and show components/vulnerabilities

I will investigate upgrading and checking if it is still there.

Thanks for any help or advice.

### Dependency-Track Version

4.7.x

### Dependency-Track Distribution

Container Image

### Database Server

PostgreSQL

### Database Server Version

Unknown, just used you Docker-Compose file

### Browser

Mozilla Firefox

### Checklist

- [X] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [X] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this defect was already reported

**Comment by nscuro** [comment]
If you're indeed running DT 4.7.x or older, I'd recommend upgrading to at least v4.9.1, which introduced support for CycloneDX v1.5.

The underlying issue here is that CycloneDX v1.5 deprecated the `metadata.tools[]` node, and added `metadata.tools.component` and `metadata.tools.services`: https://cyclonedx.org/docs/1.5/json/#tab-pane_metadata_tools_oneOf_i1

Trivy appears to be emitting BOMs using the new notation. DT versions before v4.9.x only support CycloneDX up to v1.4, that's why you're running into deserialization errors.

**Comment by nigellh** [comment]
I did wonder, thanks as always Niklas. I will do a trial upgrade on the test system and see if that fixes it. I'll close if it does.

**Comment by nigellh** [comment]
Niklas. I had badly assumed that the DT version was the 

version: '3.7'

At the top of the docker-compose file. So not exactly sure what version we were running. But I did a docker-compose pull to get the latest version and it is on bringing it up.

The error appears to be a broken database that was working fine before upgrading

```
dtrack-apiserver_1  | 2024-02-15 13:19:00,438 ERROR [UpgradeMetaProcessor] An error occurred connecting to the database.
dtrack-apiserver_1  | org.h2.jdbc.JdbcSQLNonTransientConnectionException: Unsupported database file version or invalid file header in file "/data/.dependency-track/db.mv.db" [90048-224]
```

Do I have to delete it and start from scratch or is there something I can run to fix it? I hoped it would have done the necessary upgrade.

While this is not such an issue on this one, I will have to upgrade the main one and losing the DB on that would be a big problem.

**Comment by nigellh** [comment]
Reading https://github.com/DependencyTrack/dependency-track/issues/3160

**Comment by nscuro** [comment]
It looks like you upgraded from a version earlier than v4.6.0? Can you check your production instance whether that is the case?

The version of H2 used by DT v4.5.x and earlier was end-of-life'd, and sadly the necessary H2 upgrade included breaking changes. There was no way for us to automate the migration, but we provided a guide here on how to do it manually: https://docs.dependencytrack.org/getting-started/database-support/#migrating-to-h2-v2

Arguably, it might be easier to just start over if the data is non-critical to you.

Also, if your production instance is using an external database such as Postgres instead, you will not run into this problem.

**Comment by nscuro** [comment]
The procedure mentioned in #3160 should work.

Apparently there was a breaking change that wasn't clearly stated in the changelogs, and the H2 maintainers are recommending a dump and restore: https://groups.google.com/g/h2-database/c/-HI2ubSAD-k

**Comment by nigellh** [comment]
Hi, the instructions worked fine, but still get the same error.

I used this - https://docs.dependencytrack.org/getting-started/database-support/#migrating-to-h2-v2

It appears to be what they said worked

**Comment by nigellh** [comment]
Sigh, need to read the following comments in the ticket. Will give it another go

**Comment by nigellh** [comment]
Nope if I follow his comments the jars do not exist. Will try again.

**Comment by nscuro** [comment]
> the jars do not exist

Which JARs specifically?

**Comment by nigellh** [comment]
OK turned out I had not specified the path properly (hadn't noticed the version in the path first time around). This is what I have been trying based on the page and the comments

```
wget https://repo1.maven.org/maven2/com/h2database/h2/2.1.214/h2-2.1.214.jar

java -cp h2-2.1.214.jar org.h2.tools.Script \
  -url "jdbc:h2:file:~/.dependency-track/db" \
  -user sa -password ""

tar -czf dtrack-backup.tar.gz ~/.dependency-track

rm -rf ~/.dependency-track/db.*

wget https://repo1.maven.org/maven2/com/h2database/h2/2.2.220/h2-2.2.220.jar

java -cp h2-2.2.220.jar org.h2.tools.Shell

Using the following:
    URL: jdbc:h2:~/.dependency-track/db
    Driver: org.h2.Driver (or just press Enter)
    User: sa
    Password: (Empty, just press Enter)
    Once the shell confirms the successful creation with Connected, exit the shell using the quit command

java -cp h2-2.2.220.jar org.h2.tools.RunScript \
  -url jdbc:h2:~/.dependency-track/db \
  -user sa -password "" \
  -script backup.sql \
  -options quirks_mode variable_binary
```

So the jars exist, just still getting the same error. Does this look correct? Thanks

**Comment by nscuro** [comment]
The procedure you're doing looks good to me, so not entirely sure what's wrong.

However, I just checked, and DT v4.10.1 ships with H2 v2.2.224 (reminder: we publish our own SBOM alongside releases: https://github.com/DependencyTrack/dependency-track/releases/download/4.10.1/bom.json).

![image](https://github.com/DependencyTrack/dependency-track/assets/5693141/bc531bce-bac0-425b-b3f3-0d9930143e58)

You could try repeating the process with v2.2.224 in the import step.

**Comment by nigellh** [comment]
OK, think I have it. I just logged in and ran the commands. Trouble is, where is 

~/.dependency-track/

supposed to be. I just logged in and ran things. It all worked, but I do not think it is where the database actually is. This would put it in my home directory. I did wonder, but makes sense now!

I will try and find it and fingers crossed, might get there!

**Comment by nscuro** [comment]
If you're using Docker, the database will be in `/data/.dependency-track/db.db` inside the container. `$HOME` / `~` will resolve to `/data` for the user running DT (who has the user ID `1000` per default). To illustrate:

```shell
$ docker run -d --name dt --rm dependencytrack/apiserver:4.10.1
$ docker exec -it dt /bin/bash
dtrack@c958754f2c1a:/opt/owasp/dependency-track$ id
uid=1000(dtrack) gid=1000(dtrack) groups=1000(dtrack)
dtrack@c958754f2c1a:/opt/owasp/dependency-track$ ls -la ~/.dependency-track/
total 654548
drwxr-xr-x 5 dtrack dtrack      4096 Feb 16 14:06 .
drwxrwxr-x 1 dtrack root        4096 Feb 16 14:05 ..
-rw-r--r-- 1 dtrack dtrack 670203904 Feb 16 14:07 db.mv.db
-rw-r--r-- 1 dtrack dtrack       211 Feb 16 14:05 db.trace.db
-rw-r--r-- 1 dtrack dtrack         0 Feb 16 14:05 dependency-track-audit.log
-rw-r--r-- 1 dtrack dtrack     12603 Feb 16 14:07 dependency-track.log
-rw-r--r-- 1 dtrack dtrack        36 Feb 16 14:05 id.system
drwxr-xr-x 8 dtrack dtrack      4096 Feb 16 14:05 index
drwxr-xr-x 2 dtrack dtrack      4096 Feb 16 14:05 keys
drwxr-xr-x 2 dtrack dtrack      4096 Feb 16 14:07 nist
-rw-r--r-- 1 dtrack dtrack      1148 Feb 16 14:05 server.log
```

**Comment by nigellh** [comment]
OK so this is the file

![image](https://github.com/DependencyTrack/dependency-track/assets/12110404/6b8e7748-d62e-4925-af89-ed29983779f1)

This was the command that ran without error 

sudo java -cp h2-2.1.214.jar org.h2.tools.Script \
  -url "jdbc:h2:file:/var/lib/docker/volumes/nhopper_dependency-track/_data/.dependency-trackdb" \
  -user sa -password ""

and the back up file is

![image](https://github.com/DependencyTrack/dependency-track/assets/12110404/74d03035-47b0-482a-ace1-af53c916973a)

and it has just the following in

```
-- H2 2.1.214; 
;              
CREATE USER IF NOT EXISTS "SA" SALT 'ceda277177a23ad9' HASH '0ca0d3304aebeb413e15c7d0cbdeadb430580bfd63276cf8b5a7ebd31716529b' ADMIN;     
```     

in it.

I had to run it with sudo which is not surprising as permission denied otherwise from that directory.

Clearly not extracting the original database.

Could it be the version 2.1.214 of the extraction tool?

I tried the version from the original post and that just failed with a corrupt file message.

Any suggestions?

**Comment by nigellh** [comment]
Given up, tried every version of the export script. I will just have to delete the database.

Just an FYI. I inadvertently left the DT instance running over the weekend. When I came back to the system today, it had run out of disk space. I did a docker-compose down and eventually got it all back so guessing logs? Although the docker-compose logs were not large. Ah well.

Start both systems from scratch and whinge about h2 instructions!
